### ISSUE TITLE
[Vita] Add CG shader support to piglet builds.

### ISSUE THREAD

**Issue by yamspy** [libretro/RetroArch#12414]
The piglet driver is said to only support shaders in CG format, retroarch already supports CG on other systems, please add it to vita as well.

**Comment by doctor-amaton** [comment]
I'm currently trying to do this myself. There's already a couple of "default" shaders made in CG for the backgrounds (e.g `gfx/drivers/gl_shaders/pipeline_xmb_ribbon.cg.frag.h`) which were added when support for VITAGLES was added in #11837 

I've enabled CG by adding HAVE_CG to the define flags in the Makefile, as well as updated the Pigs-in-a-Blanket project that was bundled in `deps` back then. Fixed some import issues here and there...

But I'm too dumb to actually be able to compile the darn thing correctly haha, apparently I need a libretro_vita.a (which is supossed to be a core (?) that you've already compiled before).

I'm still trying to get this to work. But if you have some idea how to get past the `make: *** No rule to make target 'libretro_vita.a', needed by 'retroarch_vita.elf'.  Stop.` I'll be happy to hear it!

**Comment by hizzlekizzle** [comment]
yes, that can be any core that you compile for vita. So, if you have a vita dev environment set up, close any core repo, build it for vita, rename the resulting *.a lib to 'libretro_vita.a' and put it in your RetroArch source tree and then compile RetroArch for vita. It will statically link against that lib and output the combined product.

**Comment by doctor-amaton** [comment]
> yes, that can be any core that you compile for vita. So, if you have a vita dev environment set up, close any core repo, build it for vita, rename the resulting *.a lib to 'libretro_vita.a' and put it in your RetroArch source tree and then compile RetroArch for vita. It will statically link against that lib and output the combined product.

Thanks! I tried that, and ended up with some undefined reference issues regarding most of the external calls in `gfx/drivers_shader/shader_gl_cg.c` (e.g `cgGLdisableClientState`) which I was trying to fix by including -lCg -lCgGl as linker options (in ubuntu, after installing the packages that provide `libCg.so` and `libCgGl.so`) but the compiler was complaining that it couldn't find them... I'm assuming I just have to update the libpath or something, but I'm not sure if all of this is going to work? like, is the PS Vita really capable of using this libraries if I'm linking some nvidia-cg-toolkit stuff that's supossed to be for amd64?

I don't even know for sure. But I've also been looking at how CG shaders for the PS Vita are being done, and the vertex / frag shaders for the XMB menu are being compiled as if they were GLSL... I've been trying to port some CG shaders and renaming them into glsl / glslp files to try and load them (PIB builds support displaying glslp not fail to apply) to no avail.

So then, we need to figure out how to link these functions in shader_gl_cg.c correctly. Maybe @SonicMastr is the right guy to help us with this!

**Comment by SonicMastr** [comment]
The reason for this error is because PIB doesn't use libCG at all. The shader compiler used in PIB is SceShaccCg, which is bound to `glCompileShader`. Because of this behavior, the correct file to use is `shader_glsl.c` and there are defines for the Vita in there.

**Comment by doctor-amaton** [comment]
Got it, thanks!

So all of the defines in `shader_glsl.c` are for importing XMB shaders, disabling snow / bokeh and providing the right CG vert and frag shaders to compile them before the menu is loaded, as well as changing `Texture` to `vTexture`. I'm only assuming we can update this file to `#if defined(VITA)` to include support for CG shader presets and types. As well as any other file that filters shaders based on the currently enabled backend (haven't found that one yet).

But before I try this out I want to make sure that you can actually apply CG shaders by changing their name and skip the extension validation. All of the ones I've tried fail to apply, maybe because of how uniforms / attributes are being passed or something, not sure. I will write a very simple, turn everything red shader for testing.

**Comment by doctor-amaton** [comment]
Ok sorry for making an additional comment, I just wanted to add that I was able to load a custom shader, by copying how the stock shader works and how are parameters defined as in and out with this weird CG / GLSL shader profile setup.

```
#if defined(VERTEX)
void main(
	float2 TexCoord,
  float2 VertexCoord,
  float4 Color,
  uniform float4x4 MVPMatrix,
  float4 out oPosition : POSITION,
  float2 out oTexCoord : TEXCOORD0,
  float4 out oColor : COLOR)
{
	oPosition = mul(float4(VertexCoord, 0.0, 1.0), MVPMatrix);
  oTexCoord = TexCoord;
  oColor = Color;
}
#elif defined(FRAGMENT)
void main(
	uniform sampler2D vTexture,
  uniform float4 bgcolor,
  float2 TexCoord : TEXCOORD0,
  float4 Color : COLOR,
  float4 out oColor : COLOR
)
{
	if (bgcolor.a > 0.0)
    oColor = bgcolor;
  else
    oColor = tex2D(vTexture, TexCoord) * Color * 0.3;
}
#endif
```
So this is a functioning shader file that can be loaded as long as you name it `*.glsl`, it will compile and it does nothing useful other than bringing the opacity down!

This, of course, rules out most of the already available cg shaders since those use the \#include prefix to bring up a bunch of compatibility stuff between CG and HLSL, which is not supported (or at least doesn't work for me) on our setup. We'd have to either rewrite the shaders themselves or implement some parsing tool to fix em before compilation.

I will edit this comment with a photograph of crt-geom working, if I can get it to work that is; as it is you can modify your presets and shaders to match this CG/GLSL blend, and it should TECHNICALLY work.

**Comment by SonicMastr** [comment]
Nice work. Happy to help with any other Piglet/ShaccCg questions

**Comment by hizzlekizzle** [comment]
that's awesome, @doctor-amaton ! I'm happy to help out with shaders, but I doubt the Vita's GPU will be strong enough for many (that's probably what's making crt-geom so slow, actually). There are still quite a few that should work just fine.

Probably the easiest way to get them will be to roll back the history of the Cg repo to before we added all of the HLSL compatibility stuff.
